### PR TITLE
Add Resource and Datasource for Netflow Exporter Policy (netflowExporterPol) (DCNE-154)

### LINKED ISSUE #1211: Add resource and datasource for netflowExporterPol (DCNE-154)
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

* Missing resource and datasource for the netflow exporter policy (class: netflowExporterPolicy).

### New or Affected Resource(s) + ACI Class(es):

* aci_netflow_exporter_policy + netflow:ExporterPol

### APIC version and APIC Platform

* V x.x.x and on-prem/cloud-aws/cloud-azure/all.

### Potential Terraform Configuration

```hcl
# Copy-paste your Terraform configurations here - for large Terraform configs,
# please use a service like Dropbox and share a link to the ZIP file. For
# security, you can also encrypt the files using our GPG public key.
```

### References

* #0000

### PR DESCRIPTION
**Pull request by gmicol**
Resolves #1211

### CODE REVIEW

File: docs/data-sources/netflow_exporter_policy.md [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+  name      = "netfow_exporter"
+}
+
+```
+
+## Schema ##
+
+### Required ###
+
+* `parent_dn` - (string) The distinguished name (DN) of the parent object, possible resources:
```
Comment (akinross) [inline docs/data-sources/netflow_exporter_policy.md:48]: Is there a default for this? like infra when tenant is not provided?

File: docs/data-sources/netflow_exporter_policy.md [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+* `name` (name) - (string) The name of the Netflow Exporter Policy object.
+
+### Read-Only ###
+
+* `id` - (string) The distinguished name (DN) of the Netflow Exporter Policy object.
+* `annotation` (annotation) - (string) The annotation of the Netflow Exporter Policy object.
+* `description` (descr) - (string) The description of the Netflow Exporter Policy object.
+* `dscp` (dscp) - (string) IP dscp value.
+* `dst_addr` (dstAddr) - (string) Remote node destination IP address.
+* `dst_port` (dstPort) - (string) Remote node destination port.
```
Comment (akinross) [inline docs/data-sources/netflow_exporter_policy.md:62]:
```suggestion
* `destination_address` (dstAddr) - (string) The destination IP address of the remote node.
* `destination_port` (dstPort) - (string) The destination port of the remote node.
```

File: docs/data-sources/netflow_exporter_policy.md [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+    - [infraInfra](https://pubhub.devnetcloud.com/media/model-doc-latest/docs/app/index.html#/objects/infraInfra/overview)
+
+* `name` (name) - (string) The name of the Netflow Exporter Policy object.
+
+### Read-Only ###
+
+* `id` - (string) The distinguished name (DN) of the Netflow Exporter Policy object.
+* `annotation` (annotation) - (string) The annotation of the Netflow Exporter Policy object.
+* `description` (descr) - (string) The description of the Netflow Exporter Policy object.
+* `dscp` (dscp) - (string) IP dscp value.
```
Comment (akinross) [inline docs/data-sources/netflow_exporter_policy.md:60]:
```suggestion
* `dscp` (dscp) - (string) The DSCP value of the Netflow Exporter Policy object.
```

File: docs/data-sources/netflow_exporter_policy.md [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+* `dst_addr` (dstAddr) - (string) Remote node destination IP address.
+* `dst_port` (dstPort) - (string) Remote node destination port.
+* `name_alias` (nameAlias) - (string) The name alias of the Netflow Exporter Policy object.
+* `owner_key` (ownerKey) - (string) The key for enabling clients to own their data for entity correlation.
+* `owner_tag` (ownerTag) - (string) A tag for enabling clients to add their own data. For example, to indicate who created this object.
+* `source_ip_type` (sourceIpType) - (string) Type of Exporter Src IP Address: Can be one of the available management IP Address for a given leaf or a custom IP Address.
+* `src_addr` (srcAddr) - (string) Source IP address.
```
Comment (akinross) [inline docs/data-sources/netflow_exporter_policy.md:67]:
```suggestion
* `source_address` (srcAddr) - (string) The source IP address.
```

File: docs/data-sources/netflow_exporter_policy.md [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+* `ver` (ver) - (string) Collector version.
```
Comment (akinross) [inline docs/data-sources/netflow_exporter_policy.md:68]: What is this collector version? Should we elaborate a bit more?
```suggestion
* `version` (ver) - (string) Collector version.
```

File: docs/data-sources/netflow_exporter_policy.md [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+* `id` - (string) The distinguished name (DN) of the Netflow Exporter Policy object.
+* `annotation` (annotation) - (string) The annotation of the Netflow Exporter Policy object.
+* `description` (descr) - (string) The description of the Netflow Exporter Policy object.
+* `dscp` (dscp) - (string) IP dscp value.
+* `dst_addr` (dstAddr) - (string) Remote node destination IP address.
+* `dst_port` (dstPort) - (string) Remote node destination port.
+* `name_alias` (nameAlias) - (string) The name alias of the Netflow Exporter Policy object.
+* `owner_key` (ownerKey) - (string) The key for enabling clients to own their data for entity correlation.
+* `owner_tag` (ownerTag) - (string) A tag for enabling clients to add their own data. For example, to indicate who created this object.
+* `source_ip_type` (sourceIpType) - (string) Type of Exporter Src IP Address: Can be one of the available management IP Address for a given leaf or a custom IP Address.
```
Comment (akinross) [inline docs/data-sources/netflow_exporter_policy.md:66]: This explanation is not really clear to me. Could you rephrase?

File: docs/resources/netflow_exporter_policy.md [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+  ]
+}
+
+```
+
+All examples for the Netflow Exporter Policy resource can be found in the [examples](https://github.com/CiscoDevNet/terraform-provider-aci/tree/master/examples/resources/aci_netflow_exporter_policy) folder.
+
+## Schema ##
+
+### Required ###
```
Comment (akinross) [inline docs/resources/netflow_exporter_policy.md:87]: same comments as datasource apply here, but should be fixed at same place so not doing twice

File: docs/resources/netflow_exporter_policy.md [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+### Optional ###
+  
+* `annotation` (annotation) - (string) The annotation of the Netflow Exporter Policy object.
+  - Default: `orchestrator:terraform`
+* `description` (descr) - (string) The description of the Netflow Exporter Policy object.
+* `dscp` (dscp) - (string) IP dscp value.
+  - Default: `CS2`
+  - Valid Values: `AF11`, `AF12`, `AF13`, `AF21`, `AF22`, `AF23`, `AF31`, `AF32`, `AF33`, `AF41`, `AF42`, `AF43`, `CS0`, `CS1`, `CS2`, `CS3`, `CS4`, `CS5`, `CS6`, `CS7`, `EF`, `VA`.
+* `dst_addr` (dstAddr) - (string) Remote node destination IP address.
+* `dst_port` (dstPort) - (string) Remote node destination port.
```
Comment (akinross) [inline docs/resources/netflow_exporter_policy.md:109]: Can only the string representation of a port be provided or also a number like http or 80?

File: gen/generator.go [hunk 9]
```diff
@@ -94,10 +94,11 @@ var templateFuncs = template.FuncMap{
 	"containsNoneAttributeValue":   ContainsNoneAttributeValue,
 	"definedInMap":                 DefinedInMap,
 	"add":                          func(val1, val2 int) int { return val1 + val2 },
-	"substract":                    func(val1, val2 int) int { return val1 - val2 },
+	"subtract":                     func(val1, val2 int) int { return val1 - val2 },
```
Comment (akinross) [inline gen/generator.go:97]: nice catch, totally missed this last review

**Review by akinross — CHANGES_REQUESTED** [review]

File: gen/generator.go (hunk 9 above)
Comment (samiib) [reply gen/generator.go:97]: Yeah I caught it in this review but not sure what happened to the PR because it got closed. 
https://github.com/CiscoDevNet/terraform-provider-aci/pull/1240#discussion_r1653734177

Comment (gmicol) [reply gen/generator.go:97]: @samiib Yeah I closed the previous PR because I really screwed up my last rebase hehe. Since I added small changes to the generator, it was easier to create a new PR for time efficiency matters.

File: gen/generator.go [hunk 10]
```diff
@@ -907,7 +917,8 @@ type Property struct {
 	DefaultValue       string
 	Versions           string
 	NamedPropertyClass string
-	ValidValues        []string
+	ValidValues        []interface{}
+	ValidValuesNames   []string
```
Comment (akinross) [inline gen/generator.go:921]: Why do we do change the ValidValues variable and then add one called ValidValuesNames, is it not easier to introduce a new variable of type []interface called ValidValuesMap?

Also we do not add the number equivalent now to the ValidValues (in your case ValidValuesNames), is this something we should?

File: gen/generator.go [hunk 11]
```diff
@@ -2224,3 +2236,33 @@ func GetValidValuesToRemove(classPkgName, propertyName string, definitions Defin
...
+					}
+				}
+
+			}
+		}
+	}
+	return false
+}
+
+func ValidValuesToMap(validValues []interface{}) map[string]string {
```
Comment (akinross) [inline gen/generator.go:2260]: Do we need this function? Or can we store this map instead of the entire validValues from meta in the property struct? This way you can refer to the map directly in the template as ValidValuesMap, and you do not have the change of ValidValues to ValidValuesNames

File: gen/templates/resource.go.tmpl [hunk 12]
```diff
@@ -15,6 +15,13 @@ import (
 	
 	"github.com/ciscoecosystem/aci-go-client/v2/client"
 	"github.com/ciscoecosystem/aci-go-client/v2/container"
+	{{- range .Properties}}
```
Comment (akinross) [inline gen/templates/resource.go.tmpl:18]: Can you try rendering without the loop, I think with a change introduced (https://github.com/CiscoDevNet/terraform-provider-aci/blob/master/main.go#L23) we might not the range and conditional (I am not sure about this and is mainly curiosity, because if this is the case we can actually simplify a lot of the import template part).

File: gen/templates/resource.go.tmpl [hunk 13]
```diff
@@ -863,10 +883,17 @@ func (r *{{.ResourceClassName}}Resource) Schema(ctx context.Context, req resourc
 					{{- if eq .Name "Annotation"}}
 				Default: stringdefault.StaticString(globalAnnotation),
 					{{- end}}
-					{{- if or .ValidValues (ne $LegacyAttributeName "")}}
+					{{- if and (lt 0 (len .ValidValues)) (lt 0 (len .Validators))}}
```
Comment (akinross) [inline gen/templates/resource.go.tmpl:886]: Can we remove (ne $LegacyAttributeName "") from the conditional without issues?

File: internal/provider/data_source_aci_endpoint_security_group.go [hunk 14]
```diff
@@ -8,6 +8,7 @@ import (
 	"context"
 	"fmt"
 
+	customtypes "github.com/CiscoDevNet/terraform-provider-aci/v2/internal/custom_types"
```
Comment (akinross) [inline internal/provider/data_source_aci_endpoint_security_group.go:11]: should we use camelCase?
```suggestion
	customTypes "github.com/CiscoDevNet/terraform-provider-aci/v2/internal/custom_types"
```

File: internal/validators/intBetweenFromString.go [hunk 15]
```diff
@@ -0,0 +1,56 @@
...
+	"strconv"
+
+	"github.com/hashicorp/terraform-plugin-framework/schema/validator"
+
+	"github.com/hashicorp/terraform-plugin-framework-validators/helpers/validatordiag"
+)
+
+var _ validator.String = inBetweenFromStringValidator{}
+
+type inBetweenFromStringValidator struct {
```
Comment (akinross) [inline internal/validators/intBetweenFromString.go:15]: Should we use PascalCase for structs?

```suggestion
type InBetweenFromStringValidator struct {
```

File: internal/validators/intBetweenFromString.go [hunk 16]
```diff
@@ -0,0 +1,56 @@
...
+	min, max int32
```
Comment (akinross) [inline internal/validators/intBetweenFromString.go:16]: Is there a reason to use int32 over int?

**Review by akinross — CHANGES_REQUESTED** [review]

File: internal/custom_types/fvRsConsIf_prio.go [hunk 17]
```diff
@@ -0,0 +1,155 @@
+package customtypes
```
[the remark this answers is not in the document]
Comment (gmicol) [reply internal/custom_types/fvRsConsIf_prio.go:1]: for the time being, let's do it per class per attribute.

File: internal/provider/data_source_aci_endpoint_security_group.go (hunk 14 above)
Comment (gmicol) [reply internal/provider/data_source_aci_endpoint_security_group.go:11]: yes, will change that

File: internal/validators/intBetweenFromString.go (hunk 16 above)
Comment (gmicol) [reply internal/validators/intBetweenFromString.go:16]: no specific reason, will change that

File: gen/generator.go [hunk 18]
```diff
@@ -2224,3 +2236,33 @@ func GetValidValuesToRemove(classPkgName, propertyName string, definitions Defin
 	}
 	return removedValidValuesSlice
 }
+
+func IsInRequiredTestValues(classPkgName, propertyName string, definitions Definitions, testType string) bool {
```
[the remark this answers is not in the document]
Comment (gmicol) [reply gen/generator.go:2240]: It enables you the ability to add to test attributes that are not required or isNaming to resource_required in testVars for specific config if you want to have specific minimum config required.

**Review by sajagana — CHANGES_REQUESTED** [review]

**Review by samiib — CHANGES_REQUESTED** [review]
LGTM

File: internal/custom_types/qosDppPol_violateMarkDscp.go [hunk 19]
```diff
@@ -0,0 +1,149 @@
...
+	priorMappedValue := QosDppPolViolateMarkDscpValueMap(v.StringValue)
+
+	newMappedValue := QosDppPolViolateMarkDscpValueMap(newValue.StringValue)
+
+	return priorMappedValue.Equal(newMappedValue), diags
+}
+
+func QosDppPolViolateMarkDscpValueMap(value basetypes.StringValue) basetypes.StringValue {
+	matchMap := map[string]string{
+		"0xffff": "unspecified",
```
[the remark this answers is not in the document]
Comment (gmicol) [reply internal/custom_types/qosDppPol_violateMarkDscp.go:117]: this is actually necessary as this semantic equality is even mentioned in the UI.
Screenshot 2024-08-21 at 11 26 36 AM

[the remark this answers is not in the document]
Comment (gmicol) [reply internal/custom_types/qosDppPol_violateMarkDscp.go:117]: This might actually be useful in the future. 
We discovered that we can actually use hexadecimal values as input for a few attributes. "flag" attributes use it as-well for RGB colors for example, but it doesn't satisfy the current conditions to receive a customtype (they don't have a validator list in their meta). 
It's too soon to know what might be the exact conditions for adding a custom type to an attribute. I think the best thing for now is to wait till we generate more resources (and start a refactor of the generator). 
This PR only focuses on semantic equality between float and string (for example TCP port values which was the original issue). 
We might need to extend the customtype generation and conditions in the future. But for now, I think it is best to stick to this solution.

File: internal/provider/resource_aci_data_plane_policing_policy.go [hunk 20]
```diff
@@ -231,14 +233,18 @@ func (r *QosDppPolResource) Schema(ctx context.Context, req resource.SchemaReque
...
+				Computed:   true,
 				PlanModifiers: []planmodifier.String{
 					stringplanmodifier.UseStateForUnknown(),
 					SetToStringNullWhenStateIsNullPlanIsUnknownDuringUpdate(),
 				},
 				Validators: []validator.String{
-					stringvalidator.OneOf("unspecified"),
+					stringvalidator.Any(
+						stringvalidator.OneOf("unspecified"),
+						validators.InBetweenFromString(0, 5.4975581376e+11),
```
[the remark this answers is not in the document]
Comment (gmicol) [reply internal/provider/resource_aci_data_plane_policing_policy.go:246]: You're right. This is also a way for the user to know that can use this value and no other string values.

File: gen/templates/custom_type.go.tmpl
[the remark this answers is not in the document]
Comment (gmicol) [reply gen/templates/custom_type.go.tmpl:1]: You're right, we should! I'll update the PR.

File: internal/custom_types/qosDppPol_violateMarkDscp.go (hunk 19 above)
[the remark this answers is not in the document]
Comment (samiib) [reply internal/custom_types/qosDppPol_violateMarkDscp.go:117]: Sounds good to me

File: gen/templates/custom_type.go.tmpl [hunk 21]
```diff
@@ -0,0 +1,151 @@
+package customTypes
```
[the remark this answers is not in the document]
Comment (samiib) [reply gen/templates/custom_type.go.tmpl:1]: What are your thoughts on  adding some extra test cases?

[the remark this answers is not in the document]
Comment (gmicol) [reply gen/templates/custom_type.go.tmpl:1]: I am adding one extra test for the customtype. 
Not sure, a negative test is useful for now, especially testing out of bounds values. We have a lot different attributes with a validator range that do not need a customtype. We just haven't implemented this yet in the resource schema. Hence we should add negative tests once it's implemented for all resources requiring a range validator and not only limited to customtype. It would be best to group these negative tests all in one PR.

File: docs/data-sources/netflow_exporter_policy.md [hunk 22]
```diff
@@ -0,0 +1,86 @@
...
+## Schema ##
+
+### Required ###
+
+* `parent_dn` - (string) The distinguished name (DN) of the parent object, possible resources:
+  - [aci_tenant](https://registry.terraform.io/providers/CiscoDevNet/aci/latest/docs/resources/tenant) ([fvTenant](https://pubhub.devnetcloud.com/media/model-doc-latest/docs/app/index.html#/objects/fvTenant/overview))
+  - The distinguished name (DN) of classes below can be used but currently there is no available resource for it:
+    - [infraInfra](https://pubhub.devnetcloud.com/media/model-doc-latest/docs/app/index.html#/objects/infraInfra/overview)
+
+  - Default: `uni/infra`
```
Comment (shrsr) [inline docs/data-sources/netflow_exporter_policy.md:53]: Did you have a discussion on whether parent_dn should be in Optional in the data source docs?

File: docs/resources/netflow_exporter_policy.md [hunk 23]
```diff
@@ -0,0 +1,189 @@
...
+
+All examples for the Netflow Exporter Policy resource can be found in the [examples](https://github.com/CiscoDevNet/terraform-provider-aci/tree/master/examples/resources/aci_netflow_exporter_policy) folder.
+
+## Schema ##
+
+### Required ###
+
+* `destination_ip_address` (dstAddr) - (string) The destination IP address of the remote node.
+* `destination_port` (dstPort) - (string) The destination port of the remote node.
+  - Valid Values: `dns`, `ftpData`, `http`, `https`, `pop3`, `rtsp`, `smtp`, `ssh`, `unspecified`.
```
Comment (akinross) [inline docs/resources/netflow_exporter_policy.md:105]: Here it seems to be missing the semantic equal port integer values?

File: docs/resources/netflow_exporter_policy.md [hunk 24]
```diff
@@ -0,0 +1,189 @@
...
+  - Default: `CS2`
+  - Valid Values:
+    * One of  `AF11`, `AF12`, `AF13`, `AF21`, `AF22`, `AF23`, `AF31`, `AF32`, `AF33`, `AF41`, `AF42`, `AF43`, `CS0`, `CS1`, `CS2`, `CS3`, `CS4`, `CS5`, `CS6`, `CS7`, `EF`, `VA`.
+    * Or a value in the range of `0` to `63`.
+* `name_alias` (nameAlias) - (string) The name alias of the Netflow Exporter Policy object.
+* `owner_key` (ownerKey) - (string) The key for enabling clients to own their data for entity correlation.
+* `owner_tag` (ownerTag) - (string) A tag for enabling clients to add their own data. For example, to indicate who created this object.
+* `source_ip_type` (sourceIpType) - (string) The type of the source IP address: It can be one of the available management IP address for a given leaf or a custom IP Address.
+  - Default: `custom-src-ip`
+  - Valid Values: `custom-src-ip`, `inband-mgmt-ip`, `oob-mgmt-ip`, `ptep`.
```
Comment (akinross) [inline docs/resources/netflow_exporter_policy.md:133]: Here it seems to be missing the semantic equal integer values?

File: docs/resources/netflow_exporter_policy.md [hunk 25]
```diff
@@ -0,0 +1,189 @@
...
+* `source_ip_address` (srcAddr) - (string) The source IP address.
+* `version` (ver) - (string) The NetFlow Exporter Version of the Netflow Exporter Policy object.
+  - Default: `v9`
+  - Valid Values: `cisco-v1`, `v5`, `v9`.
```
Comment (akinross) [inline docs/resources/netflow_exporter_policy.md:137]: Here it seems to be missing the semantic equal integer values?

File: docs/resources/netflow_exporter_policy.md [hunk 26]
```diff
@@ -0,0 +1,189 @@
...
+* `description` (descr) - (string) The description of the Netflow Exporter Policy object.
+* `qos_dscp_value` (dscp) - (string) The DSCP value of the Netflow Exporter Policy object.
+  - Default: `CS2`
+  - Valid Values:
+    * One of  `AF11`, `AF12`, `AF13`, `AF21`, `AF22`, `AF23`, `AF31`, `AF32`, `AF33`, `AF41`, `AF42`, `AF43`, `CS0`, `CS1`, `CS2`, `CS3`, `CS4`, `CS5`, `CS6`, `CS7`, `EF`, `VA`.
+    * Or a value in the range of `0` to `63`.
+* `name_alias` (nameAlias) - (string) The name alias of the Netflow Exporter Policy object.
+* `owner_key` (ownerKey) - (string) The key for enabling clients to own their data for entity correlation.
+* `owner_tag` (ownerTag) - (string) A tag for enabling clients to add their own data. For example, to indicate who created this object.
+* `source_ip_type` (sourceIpType) - (string) The type of the source IP address: It can be one of the available management IP address for a given leaf or a custom IP Address.
```
Comment (akinross) [inline docs/resources/netflow_exporter_policy.md:131]: Not sure I would add this description: `It can be one of the available management IP address for a given leaf or a custom IP Address`. Should we clean this up, and just keep the type part. The valid values already provide roughly the same information.

File: docs/resources/netflow_exporter_policy.md [hunk 27]
```diff
@@ -0,0 +1,189 @@
...
+    - Max Items: 1
+  
+
+  #### Optional ####
+    
+  * `annotation` (annotation) - (string) The annotation of the Relation To VRF object.
+      - Default: `orchestrator:terraform`
+  * `target_dn` (tDn) - (string) The distinguished name of the target.
+
+* `relation_to_epg` - (list) A list of Relation To EPG (ACI object [netflowRsExporterToEPg](https://pubhub.devnetcloud.com/media/model-doc-latest/docs/app/index.html#/objects/netflowRsExporterToEPg/overview)) pointing to  (ACI Object [fvEPg](https://pubhub.devnetcloud.com/media/model-doc-latest/docs/app/index.html#/objects/fvEPg/overview)).
```
Comment (akinross) [inline docs/resources/netflow_exporter_policy.md:149]: same as in data source

**Comment by akinross** [comment]
There are tests failing please also check CI

File: docs/resources/netflow_exporter_policy.md (hunk 23 above)
Comment (gmicol) [reply docs/resources/netflow_exporter_policy.md:105]: fixed.

File: docs/resources/netflow_exporter_policy.md (hunk 25 above)
Comment (gmicol) [reply docs/resources/netflow_exporter_policy.md:137]: There is no semantic equal integer values for `version` in the metadata.

File: docs/resources/application_epg.md [hunk 28]
```diff
@@ -266,7 +266,9 @@ All examples for the Application EPG resource can be found in the [examples](htt
   - Valid Values: `exclude`, `include`.
 * `priority` (prio) - (string) The Quality of Service (QoS) priority class ID. QoS refers to the capability of a network to provide better service to selected network traffic over various technologies. The primary goal of QoS is to provide priority including dedicated bandwidth, controlled jitter and latency (required by some real-time and interactive traffic), and improved loss characteristics. You can configure the bandwidth of each QoS level using QoS profiles.
   - Default: `unspecified`
-  - Valid Values: `level1`, `level2`, `level3`, `level4`, `level5`, `level6`, `unspecified`.
+  - Valid Values:
+    * One of  `level1`, `level2`, `level3`, `level4`, `level5`, `level6`, `unspecified`.
+    * Or a value in the range of `0` to `9`.
```
[the remark this answers is not in the document]
Comment (lhercot) [reply docs/resources/application_epg.md:271]: Well that is an APIC API question to be honest. It does accept both and it is defined with the validation that way. Maybe we need to check with Engineering on the meta being wrong. Those number are probably the queues number and the 3 missing are the reserved ones for the system.

File: docs/resources/netflow_exporter_policy.md [hunk 29]
```diff
@@ -0,0 +1,189 @@
...
+  relation_to_vrf = [
+    {
+      annotation = "annotation_1"
+      target_dn  = "uni/tn-test_tenant/ctx-test_vrf"
+    }
+  ]
+  relation_to_epg = [
+    {
+      annotation = "annotation_1"
+      target_dn  = "uni/tn-test_tenant/ap-test_ap/epg-test_epg"
```
[the remark this answers is not in the document]
Comment (gmicol) [reply docs/resources/netflow_exporter_policy.md:78]: if I reference I get the following error:
```zsh
=== RUN   TestAccResourceNetflowExporterPolWithFvTenant
    resource_aci_netflow_exporter_policy_test.go:106: Step 6/12 error: Error running pre-apply plan: exit status 1
        
        Error: Reference to undeclared resource
        
          on terraform_plugin_test.tf line 13, in resource "aci_application_epg" "test_application_epg_0":
          13:   application_profile_dn = aci_application_profile.test.id
        
        A managed resource "aci_application_profile" "test" has not been declared in
        the root module.
    panic.go:629: Error running post-test destroy, there may be dangling resources: exit status 1
        
        Error: Reference to undeclared resource
        
          on terraform_plugin_test.tf line 13, in resource "aci_application_epg" "test_application_epg_0":
          13:   application_profile_dn = aci_application_profile.test.id
        
        A managed resource "aci_application_profile" "test" has not been declared in
        the root module.
--- FAIL: TestAccResourceNetflowExporterPolWithFvTenant (30.05s)
```

[the remark this answers is not in the document]
Comment (akinross) [reply docs/resources/netflow_exporter_policy.md:78]: is the dependency present?

File: gen/testvars/fvDnsAttr.yaml [hunk 30]
```diff
@@ -22,7 +22,7 @@ resource_required:
 all:
   annotation: "annotation"
   description: "description_1"
-  filter: "filter_1"
+  filter: "test_filter"
```
[the remark this answers is not in the document]
Comment (gmicol) [reply gen/testvars/fvDnsAttr.yaml]: fix it!

File: docs/resources/netflow_exporter_policy.md (hunk 24 above)
Comment (gmicol) [reply docs/resources/netflow_exporter_policy.md:135]: This attribute/property does not satisfy all conditions to have a customtype assigned to it. It misses a validator in its metadata. Seems like it's just a fix list of valid values. It's not like Port attributes which has a fixed list of port string valid values as-well but also a range of valid numbers.

File: docs/resources/netflow_exporter_policy.md (hunk 26 above)
Comment (gmicol) [reply docs/resources/netflow_exporter_policy.md]: changed it.

**Review by shrsr — DISMISSED** [review]
LGTM

File: docs/data-sources/netflow_exporter_policy.md [hunk 31]
```diff
@@ -0,0 +1,78 @@
+---
+# Documentation generated by "gen/generator.go"; DO NOT EDIT.
+# In order to regenerate this file execute `go generate` from the repository root.
+# More details can be found in the [README](https://github.com/CiscoDevNet/terraform-provider-aci/blob/master/README.md).
+subcategory: "Tenant Policies"
+layout: "aci"
+page_title: "ACI: aci_netflow_exporter_policy"
+sidebar_current: "docs-aci-data-source-aci_netflow_exporter_policy"
+description: |-
+  Data source for ACI Netflow Exporter Policy
```
[the remark this answers is not in the document]
Comment (akinross) [reply docs/data-sources/netflow_exporter_policy.md:10]: Is this decided not to change?

File: docs/data-sources/netflow_exporter_policy.md (hunk 22 above)
Comment (akinross) [reply docs/data-sources/netflow_exporter_policy.md:53]: what is the result of this?

File: docs/resources/netflow_exporter_policy.md [hunk 32]
```diff
@@ -0,0 +1,189 @@
...
+  name_alias             = "name_alias_1"
+  owner_key              = "owner_key_1"
+  owner_tag              = "owner_tag_1"
+  source_ip_type         = "custom-src-ip"
+  source_ip_address      = "1.1.1.1/10"
+  version                = "v9"
+  relation_to_vrf = [
+    {
+      annotation = "annotation_1"
+      target_dn  = "uni/tn-test_tenant/ctx-test_vrf"
```
[the remark this answers is not in the document]
Comment (akinross) [reply docs/resources/netflow_exporter_policy.md:72]: so these we keep static?

File: docs/resources/netflow_exporter_policy.md (hunk 29 above)
[the remark this answers is not in the document]
Comment (akinross) [reply docs/resources/netflow_exporter_policy.md:78]: so these we keep static?

**Comment by akinross** [comment]
There are still a lot of open comments, are you still working on those? Could you please respond to those?

File: docs/resources/netflow_exporter_policy.md (hunk 29 above)
[the remark this answers is not in the document]
Comment (gmicol) [reply docs/resources/netflow_exporter_policy.md:78]: We can keep it static for the test but I'm currently making change to documentation so that can be overwritten in the properties.

File: docs/resources/netflow_exporter_policy.md (hunk 32 above)
[the remark this answers is not in the document]
Comment (gmicol) [reply docs/resources/netflow_exporter_policy.md:72]: We can keep it static for the test but I'm currently making change to documentation so that can be overwritten in the properties.

File: internal/provider/resource_aci_netflow_exporter_policy_test.go [hunk 33]
```diff
@@ -0,0 +1,456 @@
...
+					resource.TestCheckResourceAttr("aci_netflow_exporter_policy.test", "qos_dscp_value", "0"),
+					resource.TestCheckResourceAttr("aci_netflow_exporter_policy.test", "source_ip_address", "1.1.1.1/10"),
+				),
+			},
+		},
+	})
+}
+
+const testChildDependencyConfigNetflowExporterPol = `
+`
```
[the remark this answers is not in the document]
Comment (gmicol) [reply internal/provider/resource_aci_netflow_exporter_policy_test.go:292]: I am changing it

**Review by shrsr — APPROVED** [review]
LGTM

**Review by sajagana — APPROVED** [review]
LGTM!

File: docs/data-sources/netflow_exporter_policy.md (hunk 31 above)
[the remark this answers is not in the document]
Comment (akinross) [reply docs/data-sources/netflow_exporter_policy.md:10]: any updates on this?

File: docs/data-sources/netflow_exporter_policy.md (hunk 22 above)
Comment (akinross) [reply docs/data-sources/netflow_exporter_policy.md:53]: what is the result of this?

**Review by akinross — CHANGES_REQUESTED** [review]

File: docs/data-sources/netflow_exporter_policy.md (hunk 22 above)
Comment (gmicol) [reply docs/data-sources/netflow_exporter_policy.md:57]: Just added a new Optional Section when the parent_dn has a default value.

File: docs/data-sources/netflow_exporter_policy.md (hunk 31 above)
[the remark this answers is not in the document]
Comment (gmicol) [reply docs/data-sources/netflow_exporter_policy.md]: Just made that change.

**Review by samiib — DISMISSED** [review]
LGTM

File: gen/templates/datasource.md.tmpl [hunk 34]
```diff
@@ -46,24 +46,35 @@ Data source for ACI {{.ResourceNameAsDescription}}
 ```
 
 ## Schema ##
-
+{{ if or .HasNaming (and .HasParent (not (getDefaultValues .PkgName "parent_dn" $.Definitions)))}}
 ### Required ###
-{{ if .HasParent}}
+  {{ if and .HasParent (not (getDefaultValues .PkgName "parent_dn" $.Definitions))}}
```
Comment (sajagana) [inline gen/templates/datasource.md.tmpl:51]:
```suggestion
  {{- if and .HasParent (not (getDefaultValues .PkgName "parent_dn" $.Definitions))}}
```
Try the above change, which can reduce the unnecessary doc file changes.

File: gen/templates/testvars.yaml.tmpl [hunk 35]
```diff
@@ -132,6 +143,7 @@ targets:
   - class_name: "{{ .ClassName }}"
     target_dn: "{{ .TargetDn }}"
     target_dn_ref: "{{ .TargetDnRef }}"
+    target_dn_overwrite_docs: "{{ .TargetDnOverwriteDocs }}"
```
Comment (sajagana) [inline gen/templates/testvars.yaml.tmpl:146]:
```suggestion
    {{- if .TargetDnOverwriteDocs }}
    target_dn_overwrite_docs: "{{ .TargetDnOverwriteDocs }}"
    {{- end }}
```

File: gen/templates/testvars.yaml.tmpl [hunk 36]
```diff
@@ -149,6 +161,7 @@ child_targets:
         {{- range .ChildTestDependencies }}
   - class_name: "{{ .ClassName }}"
     target_dn: "{{ .TargetDn }}"
+    target_dn_overwrite_docs: "{{ .TargetDnOverwriteDocs }}"
```
Comment (sajagana) [inline gen/templates/testvars.yaml.tmpl:164]:
```suggestion
    {{- if .TargetDnOverwriteDocs }}
    target_dn_overwrite_docs: "{{ .TargetDnOverwriteDocs }}"
    {{- end }}
```

**Review by sajagana — CHANGES_REQUESTED** [review]

File: gen/templates/datasource.md.tmpl (hunk 34 above)
Comment (gmicol) [reply gen/templates/datasource.md.tmpl]: Fixed

File: gen/templates/testvars.yaml.tmpl (hunk 35 above)
Comment (gmicol) [reply gen/templates/testvars.yaml.tmpl:147]: Fixed

File: gen/templates/testvars.yaml.tmpl (hunk 36 above)
Comment (gmicol) [reply gen/templates/testvars.yaml.tmpl:167]: Fixed

**Review by shrsr — DISMISSED** [review]
LGTM

**Review by samiib — CHANGES_REQUESTED** [review]

File: docs/data-sources/annotation.md [hunk 37]
```diff
@@ -51,7 +51,6 @@ data "aci_annotation" "example_application_epg" {
 ## Schema ##
 
 ### Required ###
-
```
[the remark this answers is not in the document]
Comment (gmicol) [reply docs/data-sources/annotation.md]: I do agree with you. Just fixed it!

**Review by samiib — APPROVED** [review]
LGTM

**Review by sajagana — APPROVED** [review]
LGTM!

**Review by akinross — APPROVED** [review]
LGTM

**Review by anvitha-jain — APPROVED** [review]

**Review by lhercot — APPROVED** [review]
LGTM
